rbdmirror: enable rbd rados namespace mirroring (#14701)
rook/rook: Documentation/CRDs/Block-Storage/ceph-block-pool-rados-namespace-crd.md

spec:
  # The name of the CephBlockPool CR where the namespace is created.
  blockPoolName: replicapool

>   mirroring:

Reviewer: As in the other comment, let's create a separate example that is specifically for mirroring. This keeps the main example simple, and we can also mention in the separate example what the min Ceph version is for that feature.

- `blockPoolName`: The metadata name of the CephBlockPool CR where the rados namespace will be created.

- `mirroring`: Sets up mirroring of the rados namespace (requires Ceph v20 or newer)

>     - `enabled`: whether mirroring is enabled on that rados namespace (default: false)

Reviewer: > You are removing this `enabled` bool and `mirroring` is nullable, right?

Still need to remove this

    - `mode`: mirroring mode to run, possible values are "pool" or "image" (required). Refer to the [mirroring modes Ceph documentation](https://docs.ceph.com/docs/master/rbd/rbd-mirroring/#enable-mirroring) for more details
    - `remoteNamespace`: Name of the rados namespace on the peer cluster where the namespace should get mirrored.

>     - `snapshotSchedules`: schedule(s) snapshot at the **rados namespace** level. One or more schedules are supported.

Reviewer 2: Something that is probably important is to note that this is an array/list.

        - `interval`: frequency of the snapshots. The interval can be specified in days, hours, or minutes using d, h, m suffix respectively.

>         - `startTime`: optional, determines at what time the snapshot process starts, specified using the ISO 8601 time format.

Reviewer 2: Two things:

1. An example interval and start time would be helpful for users.
2. I don't see any note about required/optional on 3 of these fields, where that is noted above.

Author: >An example interval and start time would be helpful for users.

I have explained in the example, under `### Mirroring`

>I don't see any note about required/optional on 3 of these fields, where that is noted above.
All fields are optional, not sure if we do that with any where else

  pool: replicapool
  ...
```

### Mirroring

> #### Prerequisites

Reviewer: I'd suggest don't call it a prereq, it's just a necessary step
```suggestion
```

> Enable the mirroring in the default rados namespace(CephBlockPool), refer [CephBlockPool documentation](ceph-block-pool-crd.md#mirroring).

Reviewer: Let's show a simple yaml example inline of the cephblockpool. Then they can see that it's the same pool for this rados namespace. Without the example, it's not clear that they would have to set it on the `replicapool` CephBlockPool.
```suggestion
First, enable the mirroring in the default rados namespace of the CephBlockPool that contains the new rados namespace. 
```

> Configure the rados namespace CRD with mirroring spec

Reviewer: ```suggestion
Second, configure the rados namespace CRD with the mirroring enabled:
```

```yaml
apiVersion: ceph.rook.io/v1
kind: CephBlockPoolRadosNamespace
metadata:
  name: namespace-a
  namespace: rook-ceph # namespace:cluster
spec:
  # The name of the CephBlockPool CR where the namespace is created.
  blockPoolName: replicapool
  mirroring:

>     enabled: false

Reviewer: ```suggestion
    enabled: true
```

    mode: image

>     remoteNamespace: "test"

Reviewer: By default, the remote namespace might be the same? So we don't need this in the example? Otherwise, we have to explain why it is set to "test".

Author: It can be anything, it doesn thas to be same

Reviewer 3: If no remote namespace is provided, rbd-mirror daemon would mirror to a namespace with the same name as the local namespace.

Author: So currently we have made the `remote namespace` mandatory, if someone wants they can add the same name,

As their can cases:
1) Mirroring from peer1 rados namespace r1 -> Mirroring to peer2 rados namespace r2
2) Mirroring from peer1 rados namespace r1 -> Mirroring to peer2 rados namespace r1
3) Mirroring from peer1 rados namespace default -> Mirroring to peer2 rados namespace r2
4) Mirroring from peer1 rados namespace r1 -> Mirroring to peer2 rados namespace default
5) Mirroring from peer1 rados namespace default -> Mirroring to peer2 rados namespace default

For future when we support 3), 4), 5) we need to distinguish between `default` and `same name namespace` 
So when user don't specify remote namespace that will means mirror to default namespace and for same namespace user still need to specify the names

Reviewer 3: In Ceph the default namespace is specified as an empty string, so there is no need to overload on the presence of `--remote-namespace` option or similar.  I'd suggest following what RBD does:
- make `remoteNamespace` key optional
- document that `remoteNamespace: ""` would become a thing in the future

Author: Okay 👍

Reviewer: @parth-gr Let's define the type as a nullable string: `RemoteNamespace *string`. 
- When it's null (the default), we won't pass the the `--remote-namespace` flag at all
- When it's the empty string `""`, we will pass `--remote-namespace=""`, and ceph will then use the default namespace if that's the desire. 

This way, we get Ceph's current default, and they can start specifying `""` in the future when that feature is available in Ceph, with no more changes to Rook.

Author: @travisn  so should we remove this check 
`		if cephBlockPoolRadosNamespace.Spec.Mirroring.RemoteNamespace == "" { `

Plus if it is not passed we should assume it be of same name?

Author: Removed

    snapshotSchedules:
```

> #### Mirroring Spec

Reviewer: All of the spec settings should be included in a single doc section. So let's merge these settings (lines 111-117) with the spec section starting on line 49.

> - `mirroring`: Sets up mirroring of the rados namespace

Reviewer: Let's add the version requirement here as well
```suggestion
- `mirroring`: Sets up mirroring of the rados namespace (requires Ceph v20 or newer)
```

    - `enabled`: whether mirroring is enabled on that rados namespace (default: false)
    - `mode`: mirroring mode to run, possible values are "pool" or "image" (required). Refer to the [mirroring modes Ceph documentation](https://docs.ceph.com/docs/master/rbd/rbd-mirroring/#enable-mirroring) for more details

>     - `remoteNamespace`: Name of the rados namespace on the peer cluster where the namespace should get mirrored.

Reviewer: We can default to the same namespace, right?
```suggestion
    - `remoteNamespace`: Name of the rados namespace on the peer cluster where the namespace should get mirrored. The default is the same rados namespace.
```

Author: In that case we dont need to use `remotenamesapce`, we don't have default as we want it to be different

Reviewer: Per separate thread, Rook's default would be `null`, which means the param is not passed, and then Ceph's default will be the same namespace. So my suggestion still seems valid.

    - `snapshotSchedules`: schedule(s) snapshot at the **rados namespace** level. One or more schedules are supported.
        - `interval`: frequency of the snapshots. The interval can be specified in days, hours, or minutes using d, h, m suffix respectively.
        - `startTime`: optional, determines at what time the snapshot process starts, specified using the ISO 8601 time format.

!!! note

>     Rados namespace mirroring is only supported from Ceph 20.0.0.0 releases.

Reviewer: ```suggestion
    Rados namespace mirroring is only supported from Ceph v20 and newer.
```

>     And Disabling the cephblockpool namespace would require to disable mirroring on all the

Reviewer: Shouldn't this comment be in the CephBlockPool mirroring? Seems like it doesn't need to be in this doc. 
Also, notes should only contain a single concept, and this concept is different from the v20 requirement.

  pool: replicapool
  ...
```

### Mirroring

> First, enable the mirroring in the default rados namespace of the CephBlockPool that contains the new rados namespace.

Reviewer 2: What does "default rados namespace" mean? Will users know what this means?

IMO, it would be clearer to say "First, enable mirroring for the parent CephBlockPool". I understand that there is some nuance here, but 90% of users won't understand the nuance, and users who do probably/hopefully won't be tripped up.

  pool: replicapool
  ...
```

### Mirroring

First, enable mirroring for the parent CephBlockPool.

```yaml
apiVersion: ceph.rook.io/v1
kind: CephBlockPool
metadata:
  name: replicapool
  namespace: rook-ceph
spec:
  replicated:
    size: 3
  mirroring:

>     enabled: true

Reviewer: remove this

Author: We still keep the enabled with cephblockpool

Reviewer: Ah right, this is not the rados namespace

  pool: replicapool
  ...
```

### Mirroring

First, enable mirroring for the parent CephBlockPool.

```yaml
apiVersion: ceph.rook.io/v1
kind: CephBlockPool
metadata:
  name: replicapool
  namespace: rook-ceph
spec:
  replicated:
    size: 3
  mirroring:
    enabled: true
    mode: image
    # schedule(s) of snapshot
    snapshotSchedules:
      - interval: 24h # daily snapshots
        startTime: 14:00:00-05:00
```

Second, configure the rados namespace CRD with the mirroring enabled:

```yaml
apiVersion: ceph.rook.io/v1
kind: CephBlockPoolRadosNamespace
metadata:
  name: namespace-a
  namespace: rook-ceph # namespace:cluster
spec:
  # The name of the CephBlockPool CR where the namespace is created.
  blockPoolName: replicapool
  mirroring:

>     enabled: true

Reviewer: You are removing this `enabled` bool and `mirroring` is nullable, right?

    mode: image

>     snapshotSchedules:

Reviewer: Is the snapshotSchedule not required? What is the default?

Author: No default value, added some default values with `cephblockpoolvalue`

>     snapshotSchedules:

Reviewer: To merge the examples, how about adding this line?
```yaml
   remoteNamespace: namespace-a # default is the same as the local rados namespace
```

Author: actually if it is same the `remotNamesapce` doesn't require to set.

```

Another example explaining, if user wants to mirror namespace on some other remote rados namespace on the peer cluster instead of the same rados namespace, user can specify the name using `remoteNamespace`:

```yaml
apiVersion: ceph.rook.io/v1
kind: CephBlockPoolRadosNamespace
metadata:
  name: namespace-a
  namespace: rook-ceph # namespace:cluster
spec:
  # The name of the CephBlockPool CR where the namespace is created.
  blockPoolName: replicapool
  mirroring:
    enabled: true
    mode: image

>     remoteNamespace: "test" # rados namespace in the peer cluster, where namespace-a should be mirrorerd

Reviewer: If the only difference is the `remoteNamespace` in this example, I'd suggest not having the separate example.

Author: Okay
